### PR TITLE
[CMIS] Return the CDB status value for the caller to check the status…

### PR DESCRIPTION
**Pull request by xinyulin**
#### Description

This PR unifies all CDB API calls to return CDB status, allowing the caller to check the command status returned from the module and perform corresponding actions.

#### Motivation and Context
Modified the following CDB APIs to return a dictionary with CDB status as the first element and RPL result as the second element.

- query_cdb_status()
- get_module_feature()
- get_fw_management_features()
- get_fw_info()

The caller can first check the CDB status to ensure the command was processed successfully, then retrieve the data from the reply payload (RPL).

#### How Has This Been Tested?
Tested with Cisco8111 and Credo C1 cable
image

### CODE REVIEW

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py [hunk 1]
```diff
@@ -1282,7 +1282,7 @@ def get_module_fw_info(self):
             return {'status': False, 'info': "CDB Not supported", 'result': None}
 
         # get fw info (CMD 0100h)
-        rpllen, rpl_chkcode, rpl = self.cdb.get_fw_info()
+        rpllen, rpl_chkcode, rpl = self.cdb.get_fw_info()['rpl']
```
Comment (mihirpat1) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:1285]: @xinyulin Can we check for the value of `self.cdb.get_fw_info()['status']` and take appropriate action?

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py [hunk 2]
```diff
@@ -1281,7 +1284,7 @@ def get_module_fw_mgmt_feature(self, verbose = False):
         elapsedtime = time.time()-starttime
         logger.info('Get module FW upgrade features time: %.2f s\n' %elapsedtime)
         logger.info(txt)
-        return {'status': True, 'info': txt, 'feature': (startLPLsize, maxblocksize, lplonly_flag, autopaging_flag, writelength)}
+        return {'status': True, 'info': txt, 'result': (startLPLsize, maxblocksize, lplonly_flag, autopaging_flag, writelength)}
```
Comment (rqx110) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:1287]: change the field from `feature` to `result` will break the method caller!
e.g. https://github.com/sonic-net/sonic-utilities/blob/84cb00a4b2d7e8fb2bcab259367836fa11a17d0a/sfputil/main.py#L1528

Comment (xinyulin) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py:1287]: @rqx110 thanks for your review and comment, I will change it back to feature.

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py (hunk 1 above)
Comment (xinyulin) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py]: @mihirpat1 I have added the check and corresponding actions in the get_module_fw_mgmt_feature() and get_module_fw_info() functions. Could you please review them again? Thanks.

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py [hunk 3]
```diff
@@ -1277,7 +1280,7 @@ def get_module_fw_mgmt_feature(self, verbose = False):
 
         else:
             txt += 'Reply payload check code error\n'
-            return {'status': False, 'info': txt, 'result': None}
+            return {'status': False, 'info': txt, 'feature': None}
```
Comment (xinyulin) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:1283]: @mihirpat1 As suggested by rqx110, I should ensure that get_module_fw_mgmt_feature() returns a dictionary with the key 'feature' instead of 'result'; otherwise, other callers may not function properly. To ensure consistency in the output dictionary keys, I have also corrected the key to 'feature' for error cases.

File: sonic_platform_base/sonic_xcvr/api/public/cmisCDB.py [hunk 4]
```diff
@@ -155,13 +155,13 @@ def query_cdb_status(self):
...
         This QUERY Status command may be used to retrieve the password acceptance
         status and to perform a test of the CDB interface.
-        It returns the reply message of this CDB command 0000h.
+        It returns the status and reply message of this CDB command 0000h.
         '''
         cmd = bytearray(b'\x00\x00\x00\x00\x02\x00\x00\x00\x00\x10')
         cmd[133-INIT_OFFSET] = self.cdb_chkcode(cmd)
         self.write_cdb(cmd)
         status = self.cdb1_chkstatus()
-        if (status != 0x1):
```
Comment (mihirpat1) [inline sonic_platform_base/sonic_xcvr/api/public/cmisCDB.py:164]: @xinyulin Can you please revert this unrelated change to avoid increasing line history (agreed that the version without parenthesis is more pythonic though). Please do the same for similar changes in the diff.

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py (hunk 3 above)
Comment (mihirpat1) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:1283]: @xinyulin Can we log an error with the value of status here?

**Review by mihirpat1 — COMMENTED** [review]
@xinyulin Can you please test FW download + run + commit if not already tested with the changes?

File: sonic_platform_base/sonic_xcvr/api/public/cmisCDB.py (hunk 4 above)
Comment (xinyulin) [reply sonic_platform_base/sonic_xcvr/api/public/cmisCDB.py:164]: Sure, I have reverted it.

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py (hunk 3 above)
Comment (xinyulin) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py:1285]: This error will be logged with the following format in the latest commit
```python
        else:
            txt += 'Status or reply payload check code error\n'
            logger.info(txt)
            logger.info('Fail to get fw mgnt feature, cdb status: 0x{}\n'.format(status))
            return {'status': False, 'info': txt, 'feature': None}
``` 
The new commit has been tested with firmware download, run, and commit flows without any errors. Could you please review it to see if anything is missing.

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py [hunk 5]
```diff
@@ -1276,8 +1279,10 @@ def get_module_fw_mgmt_feature(self, verbose = False):
                 txt += 'Read to LPL/EPL {:#x}\n'.format(rpl[6])
 
         else:
-            txt += 'Reply payload check code error\n'
-            return {'status': False, 'info': txt, 'result': None}
+            txt += 'Status or reply payload check code error\n'
+            logger.info(txt)
+            logger.info('Fail to get fw mgnt feature, cdb status: 0x{}\n'.format(status))
```
Comment (rqx110) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:1284]:
```suggestion
            logger.info('Fail to get fw mgmt feature, cdb status: 0x{}\n'.format(status))
```

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py (hunk 3 above)
Comment (rqx110) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py:1285]: In my opinion, there is no need to log here. The method caller will make the judgment and log the result.
for example:
https://github.com/sonic-net/sonic-utilities/blob/f50587a1ff65bb489231bfe45cec805fb32dbf00/sfputil/main.py#L1526-L1531

Comment (rqx110) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py:1285]: In my opinion, there is no need to log here. The method caller will make the judgment and log the result.
for example:
https://github.com/sonic-net/sonic-utilities/blob/f50587a1ff65bb489231bfe45cec805fb32dbf00/sfputil/main.py#L1526-L1531

Comment (mihirpat1) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py:1285]: @rqx110 I think capturing the `status` and `rpl_chkcode` as part of logs will help in providing more datapoints to debug in case of an error.

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py (hunk 5 above)
Comment (mihirpat1) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py:1284]: @xinyulin Can you please log the value of `rpl_chkcode`?

Comment (xinyulin) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py]: Sure. both rpl_chkcode and check result are recorded in the new commit.

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py [hunk 6]
```diff
@@ -1276,8 +1279,10 @@ def get_module_fw_mgmt_feature(self, verbose = False):
                 txt += 'Read to LPL/EPL {:#x}\n'.format(rpl[6])
 
         else:
-            txt += 'Reply payload check code error\n'
-            return {'status': False, 'info': txt, 'result': None}
+            txt += 'Status or reply payload check code error\n'
+            logger.info(txt)
+            logger.info('Fail to get fw mgnt feature, cdb status: {:#x} rpl_chkcode: {:#x}, {:#x}\n'.format(status, self.cdb.cdb_chkcode(rpl), rpl_chkcode))
```
Comment (rqx110) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:1284]: First, `mgnt ` should `mgmt`
Second, missing ',' after `cdb status: {:#x}`
Third, the message `rpl_chkcode: {:#x}, {:#x}` is not very clear. When leaving the code, do you know what the first and second values ​​represent?

Comment (xinyulin) [reply sonic_platform_base/sonic_xcvr/api/public/cmis.py]: @rqx110 Thank you for the review. I have added more details and corrected the incorrect abbreviation.

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py [hunk 7]
```diff
@@ -1276,8 +1279,10 @@ def get_module_fw_mgmt_feature(self, verbose = False):
                 txt += 'Read to LPL/EPL {:#x}\n'.format(rpl[6])
 
         else:
-            txt += 'Reply payload check code error\n'
-            return {'status': False, 'info': txt, 'result': None}
+            txt += 'Status or reply payload check code error\n'
+            logger.info(txt)
+            logger.info('Fail to get fw mgmt feature, cdb status: {:#x}, cdb_chkcode: {:#x}, rpl_chkcode: {:#x}\n'.format(status, self.cdb.cdb_chkcode(rpl), rpl_chkcode))
```
Comment (mihirpat1) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:1284]:
```suggestion
            logger.error('Fail to get fw mgmt feature, cdb status: {:#x}, cdb_chkcode: {:#x}, rpl_chkcode: {:#x}\n'.format(status, self.cdb.cdb_chkcode(rpl), rpl_chkcode))
```

**Review by mihirpat1 — APPROVED** [review]

**Review by prgeor — APPROVED** [review]

**Comment by prgeor** [comment]
@bingwang-ms could you take this to 202405. This is for improved debugging.

**Comment by mihirpat1** [comment]
@bingwang-ms MSFT ADO - 26804772
